omniledger/service/service.go

```diff
@@ -407,63 +403,57 @@ func (s *Service) createNewBlock(scID skipchain.SkipBlockID, r *onet.Roster, cts
 	if err != nil {
 		return nil, err
 	}
-
-	s.storage.Lock()
-	pto := s.storage.PropTimeout
-	s.storage.Unlock()
-	// TODO: replace this with some kind of callback from the skipchain-service
-	log.Lvl3("Asking all nodes to update their collections")
-	replies, err := s.propagateTransactions(sb.Roster, &updateCollection{sb.Hash}, pto)
-	if err != nil {
-		log.Lvl1("Propagation-error:", err.Error())
-	}
-	if replies != len(sb.Roster.List) {
-		log.Lvlf1("%s: only got %d out of %d for block %x", s.ServerIdentity(), replies, len(sb.Roster.List), sb.Hash)
-	}
-	log.Lvl3("All nodes have the latest transactions stored")
 	return ssbReply.Latest, nil
 }
 
-// updateCollection is called once a skipblock has been stored.
-// It is called by the leader, and every node will add the
-// transactions in the block to its collection.
-func (s *Service) updateCollection(msg network.Message) {
-	uc, ok := msg.(*updateCollection)
-	if !ok {
-		log.Error("updateCollection cast failure")
-		return
+// updateCollectionCB is registered in skipchain and is called after a
+// skipblock is updated. When this function is called, it is not always after
+// the addition of a new block, but an updates to forward links, for example.
+// Hence, we need to figure out when a new block is added. This can be done by
+// looking at the latest skipblock cache from Service.state.
+func (s *Service) updateCollectionCallback(sbID skipchain.SkipBlockID) error {
```

Comment:
I see some `log.Error` - will they stay in the final code?

Reply:
I was trying to decide whether to log the errors or return them or both, which is what I'm doing now but probably shouldn't. Now I think it might be better to just remove the return value from `updateCollectionCallback` because skipchain doesn't try to handle it anyway, it only logs it. What do you think?

Reply:
I have a preference of returning errors and then printing them out if they are not nil...